#3 accdb.test

klimenkoMA opened the pull request:
Close #3 
-created DB connection
-wrote tests for the database that check CRUD operations

Commit:
Added ucanaccess to project.
Added simple entity and repository for tests. Tests added too.
Added database files to project

klimenkoMA committed:
#3 working version of the database connection, without polishing

klimenkoMA committed:
#3 changes:
- verified new DB file bot.accdb
- approved application.properties`s DB configuration
- changed annotations for UserRepositoryIntegrationTest (to @SpringBootTest)
- wrote 3 valid tests (CRUD) black box

klimenkoMA committed:
#3 changes:
- verified new DB file bot.accdb
- approved application.properties`s DB configuration
- changed annotations for UserRepositoryIntegrationTest (to @SpringBootTest)
- wrote 3 valid tests (CRUD) black box

klimenkoMA committed:
#3 changes:
- rebuild test whenCalledSave_thenCorrectWriteNewUser(), added a check for the uniqueness of the object being added to the database
- add some comments for class User
- add method List<User> findAll() to UserRepository.java

klimenkoMA committed:
#3 final (i hope ^^)) changes:
- delete training bean from main method

UserRepositoryIntegrationTest Class changes:
- create @AfterAll method DestroyTestData(@Autowired UserRepository)
- create method fail()
- wrote method whenCalledDelete_thenCorrectNumberOfUsers() and method whenCalledUpdate_thenCorrectElementChanges()
- all tests passed

spasynkov commented on src/main/resources/application.properties:
```diff
@@ -1,4 +1,9 @@
+spring.datasource.driver-class-name=net.ucanaccess.jdbc.UcanaccessDriver
+spring.jpa.properties.hibernate.dialect=net.ucanaccess.hibernate.dialect.UCanAccessDialect
+spring.jpa.hibernate.ddl-auto=none
+spring.jpa.open-in-view=false
```
давайте в этом файле всегда оставлять одну пустую строку в конце. а то оно пишет потом, что в последней строке было изменение (добавился символ `\n`)

klimenkoMA replied:
Окей, принял

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+    @Autowired
+    private UserRepository userRepository;
+
+    public void fail() {
```
а зачем этот тест? если у джейюнита есть более удобный способ завалить тест, мы ж в телеге это на днях обсуждали с Георгием

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+    private UserRepository userRepository;
+
+    public void fail() {
+        Assertions.assertEquals(0, 1);
```
так у тебя все тесты, которые вдруг завалятся - будут иметь одну и ту же ошибку, что 0 не равно 1. согласись, не очень информативная ошибка :)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+
+    @AfterAll
+    public static void destroyTestData(@Autowired UserRepository userRepos) {
+        System.out.println("Begin destroying test object`s");
```
про принтелены - забудь) их больше не существует.
все пишем логгером (посмотри в переписке мы там обсуждали как логгер подключить)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+
+
+    @AfterAll
+    public static void destroyTestData(@Autowired UserRepository userRepos) {
```
почему не `userRepository`, как и раньше?

klimenkoMA replied:
Поле класса с таким названием уже есть, мне показалось, что userRepository1 смотрится хуже, чем userRepos

spasynkov replied:
ну и ничего страшного, что существует. твоя локальная переменная в методе перекроет поле класса.
это же как в сеттерах, когда у нас есть `this.name` и просто `name`

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+    public static void destroyTestData(@Autowired UserRepository userRepos) {
+        System.out.println("Begin destroying test object`s");
+        List<User> users = userRepos.findAll();
+        for (int i = 0; i < users.size(); i++) {
```
обрати внимание, у тебя в идее for должен быть подсвечен желтым (это значит, что тут ворнинг какой-то). поставь на него курсор и нажми Alt+Enter
на ворнинги всегда обращай внимание. они чаще всего делают код лучше

klimenkoMA replied:
Тут я в трех соснах заблудился - я помню, что при работе с итератором нельзя в том же цикле удалить объект из списка. Но удаление на самом деле происходит в БД, а не в списке users. Поэтому, конечно же, нужно использовать  foreach :)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+            } catch (Exception ignored) {
+            }
+        }
+        System.out.println("All test object`s destroyed");
```
снова же, заменить на логгер

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+        System.out.println("Begin destroying test object`s");
+        List<User> users = userRepos.findAll();
+        for (int i = 0; i < users.size(); i++) {
+            try {
+                Long.parseLong(users.get(i).getName());
+                userRepos.delete(users.get(i));
+            } catch (Exception ignored) {
+            }
```
абсолютно нетривиальный блок кода.
переписать через регулярку.

klimenkoMA replied:
Вот тут ждал разворот :) Валидатор на JavaRash такую ерунду пропускал, но я понимал на тот момент, что в продакшене не может быть таких ходулей :) Просто я не изучал регулярки, и решил побыстрее сделать задание. Теперь выбора не осталось - регулярки :)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+            try {
+                Long.parseLong(users.get(i).getName());
+                userRepos.delete(users.get(i));
+            } catch (Exception ignored) {
```
если "проглатываешь" исключения - отлавливай только то, которое хочешь "проглотить", в твоем случае - NumberFormatException. это написано в джавадоках (если поставить курсор на метод `parseLong()` и нажать Ctrl+Q)

проблема в том, что если в этом блоке кода выскочит какое-то другое исключение (наследник _Exception_) - мы его точно так же "проглотим" и даже не будем знать, что у нас что-то сломалось. после того, как много дней подряд ты будешь искать почему программа работает, но не совсем так, как надо, и в итоге найдешь, что проблема была в этом - ты перестанешь так делать :)

klimenkoMA replied:
Спасибо, отличное объяснение, учту

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+        User user = new User(name, email);
+
+        //List users before save
+        List<User> userBefore = userRepository.findAll();
```
`userBefore` - не совсем удачное название переменной. я его читаю как "_пользователь ранее_". и когда я его дальше там увидел `userBefore.contains(user)` - я это прочитал как "_пользователь ранее содержит юзера? чеее?_"

лучше придумать название получше, типа, `usersListBeforeUserAdded`. не бойся длинных названий переменных. в спринге (в самой библиотеке) вообще с этим не заморачиваются.
посмотри вон [топ самых длинных названий классов в спринге](https://gist.github.com/thom-nic/2c74ed4075569da0f80b#file-output-txt) :) и это их не новички в программировании какие-то пишут.
в общем, главное - чтобы было понятно с первого взгляда на название что это и зачем

klimenkoMA replied:
Принял, переделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+        long sizeAfter;
+
+        //action
+        if (!userBefore.contains(user) & user.getName() != null || user.getEmail() != null) {
```
тут в одной строке больше всего вопросов:

- почему выбрал побитовое И вместо логического И?
- зачем проверять, что у юзера, которого ты чуть выше сам руками создал - не нулл-овые имя и имейл, если ты сам туда нулл-ы не ложил и никак с этим юзером еще не взаимодействовал? откуда там нулл-ам взяться? :)
- почему логическое ИЛИ? не будем заходить в этот иф только когда оба будут нулл-овые? типа, если одно из полей нулл - тогда все ок, можно продолжать?

klimenkoMA replied:
Побитовое И видимо от неправильного понимания механизма сравнения. Где то вычитал, что после логического И, в случае успеха, остальная часть выражения отбрасывается. Валидатор JR пропускал побитовое И, поэтому я его почти всегда его использую.
По null - решил таким образом не допустить в любом случае попадания в БД нулевых полей. Ты писал про такую проверку - я видимо места лучше не придумал, как перед записью объекта в  БД :)
Но почему то не сделал такую же проверку при update... Л-логика 
За логическое ИЛИ стыдно, грубый ляп

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+            userRepository.save(user);
+
+            //user`s after save
+            List<User> usersAfter = userRepository.findAll();
```
переименовать аналогично предыдущему списку

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+            List<User> usersAfter = userRepository.findAll();
+            sizeAfter = usersAfter.size();
+
+            //assertion
+            if (usersAfter.contains(user)) {
+                Assertions.assertEquals(sizeBefore + 1, sizeAfter);
+
+            } else {
+                System.out.println("This user not saved!");
+                fail();
+            }
```
тоже переписать через два ассершена:

- первым удостовериться, что список содержит юзера
- вторым удостовериться, что размер списка увеличился ровно на 1

7 строк твоего кода с иф-елсами заменить на две строчки ассершенов (читаемость кода повысится даже больше, чем в 3.5 раз)

klimenkoMA replied:
Принял, переделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+                System.out.println("This user not saved!");
+                fail();
+            }
+        } else {
+            System.out.println("This user already saved or has name==null or email==null!");
+            fail();
+        }
```
и снова же, про принтелены забыть, фейл теста вручную не делать.
и вообще, вынести этот блок кода перед тем, как сохранять юзера. и тоже сделать через ассершен (ассертТру, или как-то так). тогда весь предыдущий блок кода можно будет писать не в елсе, а просто как продолжение метода (потому что если ассершен не сработает - то и дальше тест выполняться уже не будет)

klimenkoMA replied:
Принял, переделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    @Test
+    public void whenCalledSave_thenCorrectWriteNewUser() {
```
только после того, как я просмотрел весь метод - я понял что он делает. так не должно быть. должно быть понятно по названию, без заглядывания в код.
предлагаю поменять на что-то типа `whenCalledSave_thenNewUserSavedCorrectly`

klimenkoMA replied:
Принял, переделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+    @Test
+    public void whenCalledDelete_thenCorrectNumberOfUsers() {
+        // taking data size and denying object
+        List<User> usersBefore = userRepository.findAll();
+        long sizeBefore = usersBefore.size();
+        User denyingUser = usersBefore.get(((int) sizeBefore - 1));
+        long userId = denyingUser.getId();
+
+        // deleting an element
+        userRepository.deleteById(userId);
+
+        //taking a new data size
+        List<User> usersAfter = userRepository.findAll();
+        long sizeAfter = usersAfter.size();
+
+        // size comparison
+        if (!usersAfter.contains(denyingUser)) {
+            Assertions.assertEquals(sizeBefore - 1, sizeAfter);
+        } else {
+            System.out.println("Deleting false!");
+            fail();
+        }
```
предлагаю переписать этот метод с учетом всех правок, что я указал выше.

но еще проблема в том, что алгоритм самого теста мне как-то не нравится. ты уверен, что эти тесты будут запускаться именно в таком порядке, в котором ты их расположил в классе? ок, может так и будет, но давай лучше перепишем более универсально, чтобы один тест не зависел от другого. потому что допустим потом кто-то другой захочет дописать еще один тестовый метод в этом классе и может получиться так, что на момент запуска этого теста - в репозитории того юзера из прошлого теста уже не будет.
в лучшем случае - тест завалится и мы это сразу увидим (если там 0 юзеров и мы попытаемся оттуда удалить кого-то)
в случае средней паршивости - наши тесты будут работать не так, как предполагалось. валиться будут одни тесты, хотя причина будет - в других. поверь, тут даже детальный дебаг не спасет (потому что дебажить же ты будешь тот тест, который заваливается, а не тот, который приводит к этому)
ну и худший случай - это если мы реально удалим тут какого-то настоящего юзера из бд.

поэтому алгоритм такой: 

1. сначала явно добавили нового юзера
2. потом явно его удалили
3. удостоверились, что:
- список уменьшился на 1 от того, которым он был после добавления
- объекта юзера нет в списке
- список ровной той длинны, какой он был и в начале теста

klimenkoMA replied:
Хорошо, переделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+
+                // element`s comparison
+                if (sizeBefore == sizeAfter) {
+                    Assertions.assertNotEquals(userBefore, userAfter);
```
ты здесь сравниваешь не юзеров, а объекты типа Optional. конечно они всегда будут разные :)

надо сначала удостовериться, что в каждом из них есть юзер, а потом уже подоставать из них юзеров и сравнивать юзеров

klimenkoMA replied:
Тут буду подробнее изучать Optional, мне его Идея выдала насильно, когда я попытался вытащить один объект из БД методом findById(). То, что я знаю об Optional на данный момент - что он упрощает код, замещая собой проверки на null. Но я не вчитывался и не разбирался глубже, поэтому для меня новость, что оказывается, из Optional нужно вытаскивать объект. Буду разбираться :)

spasynkov replied:
да там вроде не сложно. один метод проверяет наличие, другой метод достает объект

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+        // updating an element
+        User user = new User(userId, name, email);
+
+        if (!usersBefore.contains(user)) {
+            userRepository.save(user);
+            List<User> usersAfter = userRepository.findAll();
+            long sizeAfter = usersAfter.size();
+
+            if (usersAfter.contains(user)) {
+
+                //taking a new element by this id
+                Optional<User> userAfter = userRepository.findById(userId);
+
+                // element`s comparison
+                if (sizeBefore == sizeAfter) {
+                    Assertions.assertNotEquals(userBefore, userAfter);
```
трехуровневая вложенность - это плохо.
лучше в этом случае проверяй (не через ифы, а через ассерты) "негативные" сценарии. тогда если код пройдет все проверки негативных сценариев - только тогда он дойдет до твоей основной проверки неравенства юзеров.
серьезно, ты сам увидишь как твой код стал более читабельным без всех этих вложенностей

klimenkoMA replied:
Принял, переделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+
+        // taking an element by id
+        long sizeBefore = usersBefore.size();
+        long userId = usersBefore.get(((int) sizeBefore - 1)).getId();
```
снова же, надеяться, что последний элемент в бд будет именно тот, который нам надо - ненадежно.
лучше снова же сначала добавить тестовую запись, потом ее редактировать.
а удалит за тебя весь этот "мусор" тот метод, который ты пометил аннотацией `@AfterAll`

к тому же, каст к инту лучше было бы вынести на строчку выше. потому что в этой строке слишком много скобочек получается

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+
+
+    @AfterAll
+    public static void destroyTestData(@Autowired UserRepository userRepos) {
```
было бы неплохо в этом методе писать в логи сколько записей в бд было удалено по окончании работы метода

klimenkoMA replied:
Хорошо, мне нужно тогда в классе поле, которое нужно итерировать при удалении, и его значение вывести в этом методе через объект LogerFactory

spasynkov replied:
эмм, нет. внешняя переменная не нужна.
просто количество "успешных" вызовов для удаления посчитай

spasynkov reviewed (changes requested):
тесты надо немного переделать

klimenkoMA committed:
Merge branch 'dev' into #3-accdb.test

klimenkoMA committed:
#3 changes after request
- created a Logger instance for UserRepositoryIntegrationTest class
- created int remoteUserCounter
- renamed method whenCalledSave_thenNewUserSavedCorrectly()
- deleted method fail()

klimenkoMA committed:
#3 continuing rebuild tests, UserRepositoryIntegrationTest class
- if_else replaced with Assertions
- deleted System.out.println()
- commented InsuranceBot and BotConfigs because applications not started with this class

klimenkoMA committed:
#3 some changes for tests,
UserRepositoryIntegrationTest class:
- made static final LOGGER field
- added LOGGER.debug messages
- rebuild whenCalledDelete_thenCorrectNumberOfUsers method

User class:
- changed long id field to Long id field

2 tests of 3 passed

klimenkoMA committed:
#3 once the final changes for tests,
UserRepositoryIntegrationTest class:

- created deletingTestObjectsAfterMethod method
- rebuild deletingTestObjects method
- rebuild whenCalledUpdate_thenUpdatingCorrectly method

3 tests of 3 passed

spasynkov commented on src/main/java/com/example/telegrambotinsurance/entities/User.java:
```diff
@@ -0,0 +1,29 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
+@Setter
```
а зачем эти две аннотации? если аннотация `@Data` создает и геттеры и сеттеры ко всем полям

klimenkoMA replied:
Не знал, уберу :)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+	@Autowired
+	private UserRepository userRepository;
+
+	@AfterAll
+	public static void deletingTestObjects(@Autowired UserRepository userRepository) {
+		int deletingCounts = 0;
+		List<User> listOfUsers = userRepository.findAll();
+		for (User user : listOfUsers) {
+			if (user.getEmail().equals("twoBeerOrNotTwoBeer@drunkard.ru")) {
+				userRepository.delete(user);
+				deletingCounts++;
+			}
+		}
+		LOGGER.debug("Count of deleting operations  after all tests: " + deletingCounts);
+	}
+
+	@AfterEach
+	public void deletingTestObjectsAfterMethod() {
+		int deletingCounts = 0;
+		List<User> listOfUsers = userRepository.findAll();
+		for (User user : listOfUsers) {
+			if (user.getEmail().equals("twoBeerOrNotTwoBeer@drunkard.ru")) {
+				userRepository.delete(user);
+				deletingCounts++;
+			}
+		}
+		LOGGER.debug("Count of deleting operations: " + deletingCounts);
+	}
```
а чем отличаются эти два метода?

klimenkoMA replied:
Ничем, как я понял, ты попросил сделать удаление мусорных юзеров после каждого метода в классе и потом еще раз пройтись в @AfterAll

spasynkov replied:
ну да.
но я думал ты сделаешь два метода, пометишь их аннотациями нужными, а внутри каждого из них - вызовешь какой-то третий, в котором и будет находиться весь этот код. в одном месте.
или попробуешь на один и тот же метод поставить обе эти аннотации :)
но точно не дублированием кода :)

klimenkoMA replied:
теперь понял, сделаю по-другому

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		int deletingCounts = 0;
+		List<User> listOfUsers = userRepository.findAll();
+		for (User user : listOfUsers) {
+			if (user.getEmail().equals("twoBeerOrNotTwoBeer@drunkard.ru")) {
```
ок. а если нам понадобится изменить этот тестовый имейл? мне надо будет вспомнить все места в проекте, где используется этот имейл?
или может давай сделаем так, чтоб он хранился в какой-то переменной (константе), а методы бы просто обращались к ней? и тогда достаточно было бы поменять мыло только в одном месте

и давай поменяем на что-то более тривиальное :) типа `test@example.com`

klimenkoMA replied:
Принял, поменяю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+
+	@Test
+	public void whenCalledSave_thenNewUserSavedCorrectly() {
+		String name = String.valueOf(System.currentTimeMillis());
```
поскольку от идеи с регулярками отказались, и мы добились уникальности тестовым имейлом - давай тогда и имя давать какое-то понятное. например, _test user_
ну и тоже сделать так, чтобы это была общая константа для разных тестовых методов

klimenkoMA replied:
Хорошо

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		//user`s size field`s
+		listSizeBeforeSaving = listOfUsersBeforeSavingNewUser.size();
+		for (User userList : listOfUsersBeforeSavingNewUser
+		) {
```
это у тебя идея настроена так, чтобы перенос тут делать? если да - поправь, пожалуйста, чтобы в одну строку была с самим фором

klimenkoMA replied:
Хорошо

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		List<User> listOfUsersAfterSavingNewUser = userRepository.findAll();
+		listSizeAfterSaving = listOfUsersAfterSavingNewUser.size();
+		for (User userList : listOfUsersAfterSavingNewUser
+		) {
```
и тут тоже

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		) {
+			LOGGER.debug("BEFORE SAVING: " + userList);
+		}
+		LOGGER.debug("**********");
```
очень информативно (на самом деле нет)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+
+		//user`s size field`s
+		listSizeBeforeSaving = listOfUsersBeforeSavingNewUser.size();
+		for (User userList : listOfUsersBeforeSavingNewUser
+		) {
+			LOGGER.debug("BEFORE SAVING: " + userList);
+		}
+		LOGGER.debug("**********");
```
вообще не понимаю что тут происходит и зачем это все.
реально на каждого юзера в базе данных надо выводить список юзеров??

spasynkov replied:
ааа, ты там переменную назвал userList, хотя это ни разу не список, а просто юзер.
нет, названия переменных - это очень важно.

но все-равно, в логах эта информация не нужна. в продакшен базе данных могут быть сотни тысяч юзеров, а то и много миллионов. ты представляешь, что один запуск твоего теста может "сожрать" всю память, отведенную для записи логов, и из-за этого "потрутся" какие-то действительно важные записи

klimenkoMA replied:
Согласен, не нужно в логах содержимое базы, только во время написания тестов для быстрой самопроверки можно, но потом надо удалять

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		for (User userList : listOfUsersAfterSavingNewUser
+		) {
+			LOGGER.debug("AFTER SAVING: " + userList);
+		}
```
тут тоже это не надо, я считаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		LOGGER.debug("**********");
+
+		//saving new user
+		Assertions.assertFalse(listOfUsersBeforeSavingNewUser.contains(user));
```
перенеси этот ассершен повыше. там, где ты получаешь этот список. получил список - сразу проверил. логично же.

потому что проверка в блоке кода, где происходит сохранение - как-то не оч смотрится

spasynkov replied:
это решил не делать?

klimenkoMA replied:
Прошу прощения, одно из самых глупых занятий человека - игнорить ментора.
У меня две грубых ошибки:
-  переписка в гитхабе открыта в нескольких вкладках, и похоже, что я в них запутался
- я решил, что мне достаточно будет только переписать комменты, без физического перемещения строчек кода
Исправляюсь

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		Assertions.assertFalse(listOfUsersBeforeSavingUser.contains(user));
+		listSizeBeforeSaving = listOfUsersBeforeSavingUser.size();
+
+		//saving new user
```
не очень информативный коммент. по названию метода это и так понятно :)

лучше бы написал зачем ты тут сохраняешь юзера
```suggestion
		// saving new user to test update on it
```

spasynkov replied:
и это?

spasynkov replied:
второй раз отправляешь на ревью этот код без изменений... если не согласен с чем-то - ну так напиши, а не просто игнорируй

klimenkoMA replied:
Прошу прощения, исправил

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		//taking data list size
+		List<User> listOfUsersAfterSavingNewUser = userRepository.findAll();
+		listSizeBeforeDeleting = listOfUsersAfterSavingNewUser.size();
+		Assertions.assertTrue(listOfUsersAfterSavingNewUser.contains(user));
+		Assertions.assertEquals(listSizeBeforeSaving + 1, listSizeBeforeDeleting);
```
не уверен, что эти ассершены нужны. мы ведь это проверяем в методе выше...

klimenkoMA replied:
Понял, я думал каждое добавление и удаление в БД нужно проверить, изменился ли размер

spasynkov replied:
это тоже решил оставить?

spasynkov replied:
и тут тоже...

klimenkoMA replied:
Виноват, исправил

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		listSizeBeforeDeleting = listOfUsersAfterSavingNewUser.size();
+		Assertions.assertTrue(listOfUsersAfterSavingNewUser.contains(user));
+		Assertions.assertEquals(listSizeBeforeSaving + 1, listSizeBeforeDeleting);
+		for (User userList : listOfUsersAfterSavingNewUser
+		) {
+			LOGGER.debug("AFTER SAVING: " + userList);
+		}
+		LOGGER.debug("**********");
```
тоже не надо

klimenkoMA replied:
Принял, уберу

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		for (User userList : listOfUsersAfterDeletingNewUser
+		) {
+			LOGGER.debug("AFTER DELETING: " + userList);
+		}
```
и это тоже не надо

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+
+		// taking an element by id
+		userRepository.save(user);
+		userId = user.getId();
```
а зачем айдишник нужен?

klimenkoMA replied:
Я же должен знать, какую строку в БД мне нужно апдейтить. Есть уникальное мыло - но айди же будет надежнее. 
Этот айди я вставляю в юзера, который будет апдейтить существующую запись

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,181 @@
+		// create new user for update
+		String nameForUpdate = String.valueOf(System.currentTimeMillis());
+		String emailForUpdate = "twoBeerOrNotTwoBeer@drunkard.ru";
+		User userForUpdate = new User(userId, nameForUpdate, emailForUpdate);
```
аа, вижу. ты создал нового юзера...

нет, так ты проверишь, что юзер перезаписывается.
а тебе надо проверить, что юзер обновится. обновится - это когда ты существующему объекту изменил имя, вызвал метод сохранить на _этом же_ юзере - и проверил, что в базе его значение обновилось. так же и с мылом

klimenkoMA replied:
Все, понял, вытащить юзера из базы, сделать ему setName, setEmail , потом записать в базу обновленные данные

spasynkov reviewed (changes requested):
в этот раз правок уже поменьше :) в основном - поудалять лишнее

klimenkoMA committed:
#3 rebuilding  tests, User class
- deleted @Getter and @Setter annotations

klimenkoMA committed:
#3 rebuilding  tests, User class
- deleted excess LOGGER.debug() messages

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- created final static EMAIL, NAME, USER
- created delete() method
- rebuild whenCalledSave_thenNewUserSavedCorrectly() method
- rebuild whenCalledDelete_thenDeletingUserCorrectly() method

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- deleted static final USER
- created delete() method
- rebuild deletingTestObjects() method
- rebuild deletingTestObjectsAfterMethod() method

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- rebuild whenCalledUpdate_thenUpdatingCorrectly() method
- re-wrote all comments

Tests passed 3 of 3

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,135 @@
+	@AfterAll
+	public static void deletingTestObjects(@Autowired UserRepository userRepository) {delete(userRepository);}
+
+	private static void delete(@Autowired UserRepository userRepository) {
```
здесь @Autowired не нужна.
ты же сам этот метод вызываешь, а не спринг за тебя.
вот тестовые методы ты сам нигде не вызываешь. их за тебя вызывает сам спринг, поэтому там ты можешь у него попросить что-то тебе дать туда.

но метод который ты вызываешь явно, сам - спринг к нему дела не имеет. поэтому аннотация лишняя

klimenkoMA replied:
Дело в том, что @AfterAll требует статического метода, @AfterEach - не статического. Объект репозитория у нас не статический. И получается так, чтобы все три метода заработали - нужно в каждый заинжектить объект репозитория через @Autowired. И по этой же причине нельзя использовать одновременно @AfterAll и @AfterEach на одном методе

klimenkoMA replied:
С ветками вообще запутался - дев качается с сервера без изменений в файлах, мне кажется проще удалить два этих файла и скачать их заново

spasynkov replied:
это понятно. что два разных теста (один статический, второй не статический) требуют объект репозитория. тут никаких вопросов нет.

мой вопрос в том, что зачем эта аннотация в методе, который ты сам, вручную вызываешь? вот именно метод `delete()` в 27й строке тут

spasynkov replied:
что-то я смотрю ничего не изменилось по этой правке...

klimenkoMA replied:
![image](https://user-images.githubusercontent.com/71225560/97951836-8487ba00-1dac-11eb-88bf-796220f4416c.png)
Не хочет без инжекта работать. Я перепробовал много вариантов, пока не оставил этот

spasynkov replied:
ты шутишь надо мной?)
конечно он не будет так работать. ведь переменная `userRepository` не существует в области видимости этого метода.

я говорил о том, чтобы убрать саму аннотацию `@Autowired`, а не переменную вообще.
конечно без переменной оно не работает)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,135 @@
+	@Autowired
+	UserRepository userRepository;
+
+	@AfterAll
+	public static void deletingTestObjects(@Autowired UserRepository userRepository) {delete(userRepository);}
+
+	private static void delete(@Autowired UserRepository userRepository) {
+		int deletingCounts = 0;
+		List<User> listOfUsers = userRepository.findAll();
+		for(User user : listOfUsers) {
+			if(user.getEmail().equals("test@example.com")) {
+				userRepository.delete(user);
+				deletingCounts++;
+			}
+		}
+		LOGGER.debug("Count of deleting operations  after test: " + deletingCounts);
+	}
+
+	@AfterEach
+	public void deletingTestObjectsAfterMethod(@Autowired UserRepository userRepository) {delete(userRepository);}
```
а пробовал создать один метод, и на нем обе эти аннотации поставить?

klimenkoMA replied:
Да, не получилось, выше написал почему

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,135 @@
+
+		//taking list users before save
+		List<User> listOfUsersBeforeSavingUser = userRepository.findAll();
+		Assertions.assertFalse(listOfUsersBeforeSavingUser.contains(user));
```
нашел возможный баг тут. у нас ломбок по умолчанию создает метод иквалз по всем полям класса. поэтому пользователи:
```
{id=null, name="Андрей", email="andrey@gmail.com"}
{id=3210, name="Андрей", email="andrey@gmail.com"}
```
будут считаться "разными. соответственно и тут, когда ты делаешь contains - то будет проверка по иквалз, которая вернет фолс и получишь фолс.
хотя логически мы понимаем, что это один и тот же юзер.

поэтому я предлагаю исключить айдишник в классе Юзер из иквалза. и поскольку иквалз мы не сами писали, а нам его ломбок сгенерировал - то почитай в интернете какую аннотацию и куда надо поставить для того, чтобы исключить из иквалз какое-то поле :)

klimenkoMA replied:
Если я тебя правильно понял - я этот момент описывал в телеге. 
![image](https://user-images.githubusercontent.com/71225560/97797513-cd196900-1c2e-11eb-8c93-82a0697ea666.png)
В тот момент, когда произошла запись юзера с нулевым id в БД - id перестает быть нулевым и в БД и в объекте, с которым продолжает работать программа. Поэтому и все ассерты, проверяющие наличие нашего юзера в списке - срабатывают без фейлов. Однако, ты гораздо опытнее меня и знаешь в 100500 раз больше - возможно, в продакшене что-то может поменяться. И тогда лучше обезопасить тестирование исключением поля id из сравнения

spasynkov replied:
ну тут такой момент, как мы считаем, одинаковые или разные такие объекты? один и тот же юзер, или разные юзеры?
```
{id=null, name="Андрей", email="andrey@gmail.com"}
{id=3210, name="Андрей", email="andrey@gmail.com"}
```

или вот такие
```
{id=1234, name="Андрей", email="andrey@gmail.com"}
{id=3210, name="Андрей", email="andrey@gmail.com"}
```

это разные пользователи или одинаковые в нашем понимании? :)

если в бд есть пользователь `{id=3210, name="Андрей", email="andrey@gmail.com"}`
и я хочу добавить пользователя `{id=null, name="Андрей", email="andrey@gmail.com"}`
я проверяю через `contains()` - получаю `false` (потому что `equals()` вернет `false`)
добавляю юзера
он идет логиниться. в какой из своих профилей он попадет? а при следующем логине? :) неизвестно
а если предположить, что на новом аккаунте он оплатил дорогущую подписку, а на старом нет - вот тебе недовольный клиент, гневные отзывы в интернете и может даже заявки в суд (по вине программиста. потому что программист посчитал, что это разные объекты юзеров)

spasynkov replied:
то-есть, тут вопрос не в том, что во время записи в бд юзеру присваивается айдишник. это правильно
а в том, что джава считает объект с айдишником и без - разными объектами по своей сути

klimenkoMA replied:
Теперь понял, спасибо :) Исправлю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/entities/User.java:
```diff
@@ -0,0 +1,27 @@
+
+import lombok.*;
+
+import java.util.List;
```
а зачем здесь этот импорт? у тебя ж нигде не используется список никакой

klimenkoMA replied:
![image](https://user-images.githubusercontent.com/71225560/97808856-5bffa300-1c7a-11eb-86d1-eb03f0371b45.png)
вот тут используется :)

spasynkov replied:
это хорошо, но я тебе про класс User, а ты мне - про тесты :)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,135 @@
+
+	@Test
+	public void whenCalledUpdate_thenUpdatingCorrectly() {
+		String name = String.valueOf(System.currentTimeMillis());
+		String email = "masterOfUniverse@deathstar.empire";
```
а тут ничего не надо было поменять?

spasynkov replied:
ааа, или это такая задумка? сначала создать юзера, а потом сделать ему апдейт?

klimenkoMA replied:
Да, мы же все равно перезапишем поля этого юзера под статические , поэтому неважно, какие поля использовать до апдейта

spasynkov replied:
ага, ок

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,135 @@
+		//saving new user
+		userRepository.save(user);
+
+		//checking the absence of a new user in the database and checking the database size change
```
а отсутствие чего ты тут проверяешь? а то я не вижу что-то такой проверки...

klimenkoMA replied:
Исправил

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,135 @@
+		user.setName(NAME);
+		user.setEmail(EMAIL);
+
+		//checking the absence of a new user in the database
+		Assertions.assertFalse(listOfUsersAfterSavingNewUser.contains(user));
```
а зачем эта проверка?
опустим она не прошла и тест завалился. нам это о чем-то говорит? я просто не понимаю что ты проверяешь тут :)

klimenkoMA replied:
Проверяю, нет ли такого юзера в базе. Если завалится  - значит есть юзер, и это будет значить, что при тесте не почистилась база почему-то, либо чьи-то шуточки :)

spasynkov replied:
ну тогда вот это и надо было добавить в комментарий.
то, что делает метод contains - понятно и так. а вот зачем - лучше уточнить :)

klimenkoMA replied:
Окей, добавлю :)

spasynkov replied:
тоже, написал что добавишь, но ничего не изменилось... передумал? почему не отписал тогда?

klimenkoMA replied:
Не дописал, прочел невнимательно, прошу прощения
Исправил

spasynkov reviewed (changes requested):
пару моментов еще осталось неясных...
да и ветки до сих пор конфликтуют...

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- moved the variable deletingCounts out of the loop in delete() method
- re-wrote some comments

klimenkoMA committed:
#3 rebuilding  tests, User class
- added @EqualsAndHashCode(exclude = "id")

UserRepositoryIntegrationTest class
- added static int deletingCounts

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- changed some comments

spasynkov reviewed (changes requested):
в основном осталось то, что я писал раньше, но ты проигнорировал это.
пока мы по этим моментам не придем к общему мнению - код в дев не попадет :)

klimenkoMA committed:
#3 rebuilding  tests, User class
- deleted import *.*.List

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- redesigned the placement of Assertions checks for logical relationships

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- redesigned the placement of Assertions checks for logical relationships
 in whenCalledDelete_thenDeletingUserCorrectly() method

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- redesigned the placement of Assertions checks for logical relationships
 in whenCalledUpdate_thenUpdatingCorrectly() method

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- added some changes in comments

spasynkov commented on src/test/java/com/example/telegrambotinsurance/repositories/UserRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,148 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(UserRepositoryIntegrationTest.class);
+	private static final String NAME = "testUser";
+	private static final String EMAIL = "test@example.com";
+	private static int deletingCounts = 0;
```
я смотрю ты сделал эту переменную статической?
а зачем?
программу запускал, логи смотрел?
![1](https://user-images.githubusercontent.com/14253682/98048676-df65f380-1e36-11eb-8dec-df8cd64f5288.png)
тебя не смущает, что там судя по логам после каждого теста из базы удаляется все больше и больше записей? это что получается, у тебя с каждым тестом все больше и больше говна в бд пишется? или все-таки проблема со счетчиком?

klimenkoMA replied:
![image](https://user-images.githubusercontent.com/71225560/98076601-9c7c3e00-1e7f-11eb-8a40-1528a833b42a.png)
На этот пост я ориентировался, почему я решил, что нужно количество вызовов метода delete() моего класса посчитать, а не количество реально удаленных записей из БД - я не знаю. 
У меня был реализован правильный вариант подсчета удаленных записей - я его в последнем пуше переписал вот таким образом
Исправлю

spasynkov reviewed (changes requested):
верни переменную, но без аннотации
и разберись со счетчиком

klimenkoMA committed:
#3 rebuilding  tests, UserRepositoryIntegrationTest class
- deleted @Autowired annotation from delete() method
- moved variable deletingCounts inside the loop in delete() method

klimenkoMA committed:
#3 try to remove the conflict with dev branch

spasynkov reviewed (approved):
вроде все ок

spasynkov committed:
Merge branch 'dev' into #3-accdb.test
